processor/spanmetricsprocessor/processor.go

```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		mLatency.IntHistogram().DataPoints().Append(dpLatency)
+		mLatency.IntHistogram().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mLatency)
+	}
+	return nil
+}
+
+// collectCallMetrics collects the raw call count metrics, writing the data
+// into the given instrumentation library metrics.
+func (p *processorImp) collectCallMetrics(ilm *pdata.InstrumentationLibraryMetrics) error {
+	for key := range p.callSum {
+		dpCalls := pdata.NewIntDataPoint()
+		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
+		dpCalls.SetValue(p.callSum[key])
+
+		kmap := make(map[string]string)
+		if err := p.jsonSerder.Unmarshal([]byte(key), &kmap); err != nil {
+			return err
+		}
+		dpCalls.LabelsMap().InitFromMap(kmap)
+
+		mCalls := pdata.NewMetric()
+		mCalls.SetDataType(pdata.MetricDataTypeIntSum)
+		mCalls.SetName("calls")
+		mCalls.IntSum().DataPoints().Append(dpCalls)
+		mCalls.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mCalls)
+	}
+	return nil
 }
 
 // aggregateMetrics aggregates the raw metrics from the input trace data.
 // Each metric is identified by a key that is built from the service name
 // and span metadata such as operation, kind, status_code and any additional
 // dimensions the user has configured.
-func (p *processorImp) aggregateMetrics(traces pdata.Traces) {
-	// TODO: Add implementation
+func (p *processorImp) aggregateMetrics(traces pdata.Traces) error {
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		rspans := traces.ResourceSpans().At(i)
+		r := rspans.Resource()
+
+		attr, ok := r.Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+		serviceName := attr.StringVal()
+		if err := p.aggregateMetricsForServiceSpans(rspans, serviceName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *processorImp) aggregateMetricsForServiceSpans(rspans pdata.ResourceSpans, serviceName string) error {
+	ilsSlice := rspans.InstrumentationLibrarySpans()
+	for j := 0; j < ilsSlice.Len(); j++ {
+		ils := ilsSlice.At(j)
+		spans := ils.Spans()
+		for k := 0; k < spans.Len(); k++ {
+			span := spans.At(k)
+			if err := p.aggregateMetricsForSpan(serviceName, span); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Span) error {
+	key, err := p.buildKey(serviceName, span)
+	if err != nil {
+		return err
+	}
+
+	p.lock.Lock()
+	p.updateCallMetrics(key)
+	p.lock.Unlock()
+
+	latency := float64(span.EndTime()-span.StartTime()) / float64(time.Millisecond.Nanoseconds())
+
+	// Binary search to find the latency bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latency)
+
+	p.lock.Lock()
+	p.updateLatencyMetrics(key, latency, index)
+	p.lock.Unlock()
+	return nil
+}
+
+// updateCallMetrics increments the call count for the given metric key.
+func (p *processorImp) updateCallMetrics(key string) {
+	p.callSum[key]++
```

Comment:
Looks like we never retire anything from `callSum`, `latencySum`, etc maps. Will they grow infinitely? Should we reset them periodically with fresh start times?

Reply:
Sorry, I was not clear. I did not mean growing the value of a particular sum stored in callSum entry. I mean growing the number of entires in the callSum map, so that the map become larger and larger as we observe new keys.

Reply:
As for resetting the value I think that's the expected behavior for cumulative metrics, see [here] (https://github.com/open-telemetry/opentelemetry-proto/blob/b9f9b7e6d0ecbf00784a02e1b17fd5cc00ee7d73/opentelemetry/proto/metrics/v1/metrics.proto#L269):

>  // CUMULATIVE is an AggregationTemporality for a metric aggregator which
>  // reports changes since a fixed start time. This means that current values
>  // of a CUMULATIVE metric depend on all previous measurements since the
>  // start time. Because of this, the sender is required to retain this state
>  // in some form. **If this state is lost or invalidated, the CUMULATIVE metric**
>  // **values MUST be reset and a new fixed start time following the last**
>  // **reported measurement time sent MUST be used.**

I may be wrong, but that's my understanding of how you can start the counters from 0. @bogdandrutu @jmacd can you confirm?

Reply:
> I mean growing the number of entires in the callSum map, so that the map become larger and larger as we observe new keys.

Okay, yes indeed the map keys will continue to grow. I'll look into resetting the maps along with resetting start times.

Reply:
@tigrannajaryan I wrote a small test where on every 10K spans received I try resetting my raw metrics maps and startTime like so:
```
func (p *processorImp) resetRawMetrics() {
	p.counter++
	if p.counter > 10000 {
		p.callSum = make(map[string]int64)
		p.latencySum = make(map[string]float64)
		p.latencyCount = make(map[string]uint64)
		p.latencyBucketCounts = make(map[string][]uint64)
		p.startTime = time.Now()
		p.counter = 0
	}
}
...
		dpLatency := pdata.NewIntHistogramDataPoint()
		dpLatency.SetStartTime(pdata.TimeToUnixNano(p.startTime))
		dpLatency.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
		dpLatency.SetExplicitBounds(p.latencyBounds)
		dpLatency.SetBucketCounts(p.latencyBucketCounts[key])
...
		dpCalls := pdata.NewIntDataPoint()
		dpCalls.SetStartTime(pdata.TimeToUnixNano(p.startTime))
		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
		dpCalls.SetValue(p.callSum[key])
...
```

However, my metric numbers are now jumping around, with some metrics also missing then reappearing again when I query my prometheus server's `/metrics` endpoint. Is there something I'm missing or doing wrong that would prevent Prometheus from stitch together disjoint timeseries?

Reply:
I don't know much about Prometheus to help. I am fine if you keep the ever-increasing metrics for now and solve this in a future PR.

Reply:
Okay, I've created a follow-up Issue to address this at a later time: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2179.